tls/s2n_send.c

```diff
@@ -97,40 +94,15 @@ int s2n_flush(struct s2n_connection *conn, s2n_blocked_status *blocked)
         }
         conn->wire_bytes_out += w;
     }
-
-    if (conn->write_closing) {
-        conn->write_closed = 1;
-    }
     POSIX_GUARD(s2n_stuffer_rewrite(&conn->out));
 
-    /* If there's an alert pending out, send that */
-    if (s2n_stuffer_data_available(&conn->reader_alert_out) == 2) {
-        struct s2n_blob alert = { 0 };
-        alert.data = conn->reader_alert_out.blob.data;
-        alert.size = 2;
-        POSIX_GUARD_RESULT(s2n_record_write(conn, TLS_ALERT, &alert));
-        POSIX_GUARD(s2n_stuffer_rewrite(&conn->reader_alert_out));
-        POSIX_GUARD_RESULT(s2n_alerts_close_if_fatal(conn, &alert));
-
-        /* Actually write it ... */
-        goto WRITE;
-    }
-
-    /* Do the same for writer driven alerts */
-    if (s2n_stuffer_data_available(&conn->writer_alert_out) == 2) {
-        struct s2n_blob alert = { 0 };
-        alert.data = conn->writer_alert_out.blob.data;
-        alert.size = 2;
-        POSIX_GUARD_RESULT(s2n_record_write(conn, TLS_ALERT, &alert));
-        POSIX_GUARD(s2n_stuffer_rewrite(&conn->writer_alert_out));
-        POSIX_GUARD_RESULT(s2n_alerts_close_if_fatal(conn, &alert));
-
-        /* Actually write it ... */
-        goto WRITE;
+    if (conn->reader_warning_out) {
+        POSIX_GUARD_RESULT(s2n_alerts_write_warning(conn));
+        conn->reader_warning_out = 0;
+        POSIX_GUARD(s2n_flush(conn, blocked));
```

Comment:
Recursion?? You found an appropriate use of recursion??

Reply:
I mean... I found *a* use of recursion ;) It's a little cleaner than the previous goto.